firedrake/slate/slate.py

```diff
@@ -0,0 +1,673 @@
+"""SLATE is a symbolic language defining a framework for
+performing linear algebra operations on finite element
+tensors. It is similar in principle to most linear algebra
+libraries in notation.
+
+The design of SLATE was heavily influenced by UFL, and
+utilizes much of UFL's functionality for FEM-specific form
+manipulation.
+
+Unlike UFL, however, once forms are assembled into SLATE
+`Tensor` objects, one can utilize the operations defined
+in SLATE to express complicated linear algebra operations.
+(Such as the Schur-complement reduction of a block-matrix
+system)
+
+All SLATE expressions are handled by a specialized linear
+algebra compiler, which interprets SLATE expressions and
+produces C++ kernel functions to be executed within the
+Firedrake architecture.
+"""
+
+import operator
+
+from ufl.algorithms.map_integrands import map_integrand_dags
+from ufl.algorithms.multifunction import MultiFunction
+from ufl.domain import join_domains, sort_domains
+
+
+__all__ = ['Tensor', 'Scalar', 'Vector', 'Matrix',
+           'Inverse', 'Transpose', 'UnaryOp', 'Negative', 'Positive',
+           'BinaryOp', 'TensorAdd', 'TensorSub', 'TensorMul']
+
+
+class CheckRestrictions(MultiFunction):
+    """UFL MultiFunction for enforcing cell-wise integrals to contain
+    only positive restrictions."""
+
+    expr = MultiFunction.reuse_if_untouched
+
+    def negative_restrictions(self, o):
+        raise ValueError("Cell-wise integrals must contain only positive restrictions.")
+
+
+class Tensor(object):
+    """An abstract representation of a finite element tensor in SLATE.
+    All derived tensors, `Scalar`, `Vector`, and `Matrix`, as well as all
+    `BinaryOp` and `UnaryOp` objects derive from this abstract base class.
+    """
+    # Initializing tensor id class variable for output purposes
+    id = 0
+
+    def __init__(self):
+        """Constructor for the Tensor class."""
+        self.id = Tensor.id
+        Tensor.id += 1
+
+        # Compute Tensor shape information
+        shape = []
+        shapes = {}
+
+        for i, arg in enumerate(self.arguments()):
```

Comment:
Calling overridden methods from the base class constructor is considered bad style (at least), since you can normally expect that calling a method is safe after the object is fully constructed. Since the base class is normally initialised before the derived class (the contrary is even illegal in C++ and Java, while in Python is used to cover up poor design), you should assume that the derived class isn't initalised yet.

Luckily, you don't need to do this here in the constructor. It would be clearer if you expressed `shapes`, `shape` and `rank` as properties instead.